client.go

```diff
@@ -143,27 +178,34 @@ func (c *Client) Do(req *Request) (*http.Response, error) {
 
 		// Attempt the request
 		resp, err := c.HTTPClient.Do(req.Request)
+
+		// check if we should continue with retries
+		checkErr := c.CheckRetry(resp, err)
+
 		if err != nil {
 			c.Logger.Printf("[ERR] %s %s request failed: %v", req.Method, req.URL, err)
-			goto RETRY
+		} else {
+			// call this here to maintain the behavior of logging all requests,
```

Comment:
This still looks pretty much the same. We guard this off in the else clause here, because we can't call `ResponseLogHook` if there was an error.  I agree this is little awkward, but it ensures the existing behavior was maintained where the logger is called even when we don't retry. I would normally have just dropped this down into the retry section where we drain the body.